br/pkg/lightning/checkpoints/checkpoints.go

```diff
@@ -940,63 +948,100 @@ func (cpdb *MySQLCheckpointsDB) Update(checkpointDiffs map[string]*TableCheckpoi
-		}
-		// FIXME: patch for empty map may need initialize manually, because currently
-		// FIXME: a map of zero size -> marshall -> unmarshall -> become nil, see checkpoint_test.go
-		if cpdb.checkpoints.Checkpoints == nil {
-			cpdb.checkpoints.Checkpoints = map[string]*checkpointspb.TableCheckpointModel{}
-		}
-		for _, table := range cpdb.checkpoints.Checkpoints {
-			if table.Engines == nil {
-				table.Engines = map[int32]*checkpointspb.EngineCheckpointModel{}
-			}
-			for _, engine := range table.Engines {
-				if engine.Chunks == nil {
-					engine.Chunks = map[string]*checkpointspb.ChunkCheckpointModel{}
-				}
-			}
-		}
-	} else {
+
+	// init ExternalStorage
+	s, fileName, err := createExstorageByCompletePath(ctx, path)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	cpdb.ctx = ctx
+	cpdb.fileName = fileName
+	cpdb.exStorage = s
+
+	exist, err := cpdb.exStorage.FileExists(ctx, fileName)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if !exist {
 		log.L().Info("open checkpoint file failed, going to create a new one",
 			zap.String("path", path),
 			log.ShortError(err),
 		)
+		return cpdb, nil
+	}
+	content, err := cpdb.exStorage.ReadFile(ctx, fileName)
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
-	return cpdb
+	err = cpdb.checkpoints.Unmarshal(content)
+	if err != nil {
+		log.L().Error("checkpoint file is broken", zap.String("path", path), zap.Error(err))
+	}
+	// FIXME: patch for empty map may need initialize manually, because currently
+	// FIXME: a map of zero size -> marshall -> unmarshall -> become nil, see checkpoint_test.go
+	if cpdb.checkpoints.Checkpoints == nil {
+		cpdb.checkpoints.Checkpoints = map[string]*checkpointspb.TableCheckpointModel{}
+	}
+	for _, table := range cpdb.checkpoints.Checkpoints {
+		if table.Engines == nil {
+			table.Engines = map[int32]*checkpointspb.EngineCheckpointModel{}
+		}
+		for _, engine := range table.Engines {
+			if engine.Chunks == nil {
+				engine.Chunks = map[string]*checkpointspb.ChunkCheckpointModel{}
+			}
+		}
+	}
+	return cpdb, nil
 }
 
-func (cpdb *FileCheckpointsDB) save() error {
-	serialized, err := cpdb.checkpoints.Marshal()
+// createExstorageByCompletePath create ExternalStorage by completePath and return fileName.
+func createExstorageByCompletePath(ctx context.Context, completePath string) (storage.ExternalStorage, string, error) {
+	purl, err := storage.ParseRawURL(completePath)
 	if err != nil {
-		return errors.Trace(err)
+		return nil, "", errors.Trace(err)
 	}
-	// because `os.WriteFile` is not atomic, directly write into it may reset the file
-	// to an empty file if write is not finished.
-	tmpPath := cpdb.path + ".tmp"
-	if err := os.WriteFile(tmpPath, serialized, 0o600); err != nil {
-		return errors.Trace(err)
+
+	fileName := path.Base(purl.Path)
+	fileDir := path.Dir(purl.Path)
+
+	purl.Path = fileDir
+	newURL, err := url.QueryUnescape(purl.String())
```

Comment:
Why unescape the query? I think url parser should do this implicitly.

Reply:
ok, i will add some unit test like dm.